### PR TITLE
Rationalize and expand testing model

### PR DESCRIPTION
**Pull request by RichardBruskiewich**
- Added **OneHopTests** and **StandardsValidation** to **TestObjectiveEnum**
-  Refactored **`test_env`** and **`test_runner_settings`** upwards to **TestRunSession**
- Embed a multivalued **`test_run_parameters`** tag=value' slot in **TestRunSession** to allow for global configuration of TestRunner
- Renamed **`runner_settings`** to **`test_runner_settings`** and move upwards to **TestEntity**, for fine grained test-data driven local configuration of a **TestRunner**
- Removed unneeded **TestCase** subclasses

### CODE REVIEW

File: project/graphql/translator_testing_model.graphql [hunk 1]
```diff
@@ -324,22 +265,17 @@ type TestResultPKSet
...
-  }
-
 type TestRunSession
   {
     id: Uriorcurie!
     name: String
     description: String
     tags: [String]
+    testRunnerSettings: [String]
+    components: [ComponentEnum]
```
[the remark this answers is not in the document]
Comment (maximusunc) [reply project/graphql/translator_testing_model.graphql:275]: A use case could be TRAPI validation. We would hit all levels, KPs, ARAs, ARS, make sure that none of them are breaking the standard.

File: src/translator_testing_model/datamodel/pydanticmodel.py [hunk 2]
```diff
@@ -53,6 +53,10 @@ class TestObjectiveEnum(str, Enum):
     BenchmarkTest = "BenchmarkTest"
     # Quantitative test
     QuantitativeTest = "QuantitativeTest"
+    # One Hop Tests of knowledge graph navigation
+    OneHopTests = "OneHopTests"
```
Comment (maximusunc) [inline src/translator_testing_model/datamodel/pydanticmodel.py:57]: Should this be plural?

File: src/translator_testing_model/datamodel/pydanticmodel.py [hunk 3]
```diff
@@ -53,6 +53,10 @@ class TestObjectiveEnum(str, Enum):
...
+    # TRAPI and Biolink Model ("reasoner-validator") validation
+    StandardsValidation = "StandardsValidation"
```
[the remark this answers is not in the document]
Comment (maximusunc) [reply src/translator_testing_model/datamodel/pydanticmodel.py:59]: Or remove "Test" from all?

File: src/translator_testing_model/datamodel/pydanticmodel.py [hunk 4]
```diff
@@ -184,19 +188,12 @@ class TrapiTemplateEnum(str, Enum):
...
-    
-    bte = "bte"
-    
-    improving = "improving"
+class ComponentEnum(str):
+    """
+    Translator components are identified by their InfoRes identifiers.
+    """
     
+    dummy = "dummy"
```
Comment (maximusunc) [inline src/translator_testing_model/datamodel/pydanticmodel.py:196]: What's the goal here?

File: src/translator_testing_model/datamodel/pydanticmodel.py [hunk 5]
```diff
@@ -558,6 +487,7 @@ class StandardsComplianceTestSuite(TestSuite):
     name: Optional[str] = Field(None, description="""A human-readable name for a Test Entity""")
     description: Optional[str] = Field(None, description="""A human-readable description for a Test Entity""")
     tags: Optional[List[str]] = Field(default_factory=list, description="""A human-readable tags for categorical memberships of a TestEntity (preferably a URI or CURIE). Typically used to aggregate instances of TestEntity into formally typed or ad hoc lists.""")
+    test_runner_settings: Optional[List[str]] = Field(default_factory=list, description="""Scalar parameters for the TestRunner processing a given TestEntity.""")
     
 
 class OneHopTestSuite(TestSuite):
```
Comment (maximusunc) [inline src/translator_testing_model/datamodel/pydanticmodel.py:493]: Does the OneHopTestSuite not need a TestCase class?

Comment (maximusunc) [reply src/translator_testing_model/datamodel/pydanticmodel.py:493]: Same for TRAPIValidation

Comment (RichardBruskiewich) [reply src/translator_testing_model/datamodel/pydanticmodel.py:493]: @maximusunc, isn't that covered by the class attribute:

```
 test_cases: Optional[Dict[str, TestCase]] = Field(default_factory=dict, description="""List of explicitly enumerated Test Cases.""")
```

That being said, I'm less convinced that these subclasses of **TestSuite** - i.e.  **AcceptanceTestSuite**, **BenchmarkTestSuite** (this one looks weird...), **StandardsComplianceTestSuite** and **OneHopTestSuite** - are even necessary (anymore, if they ever were) and could perhaps be removed from the model.  @sierra-moxon .. any thoughts about this?

File: src/translator_testing_model/datamodel/pydanticmodel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RichardBruskiewich) [reply src/translator_testing_model/datamodel/pydanticmodel.py]: OK. Done

File: src/translator_testing_model/datamodel/pydanticmodel.py (hunk 2 above)
Comment (RichardBruskiewich) [reply src/translator_testing_model/datamodel/pydanticmodel.py]: I guess not, for consistency. I'll change that. Done

File: project/graphql/translator_testing_model.graphql (hunk 1 above)
[the remark this answers is not in the document]
Comment (RichardBruskiewich) [reply project/graphql/translator_testing_model.graphql:275]: Hi @maximusunc, **`components`** are a **`List[str]`** of **infores** identifiers of target **`components`**. The implicit assumption that was made (perhaps incorrect, I guess) was that _all_ of the **`test_entities`** (test data - **TestAssets/Cases/Suites**) of a given **TestRunSession** will be run against _all_ of the specified **`components`**. If a TestRunSession is a heterogeneous mix of testing targets for which the data is not mapped one-to-one, then of course, this assumption is violated.  I guess it just depends how one decides to cut the cake.

[the remark this answers is not in the document]
Comment (RichardBruskiewich) [reply project/graphql/translator_testing_model.graphql:275]: As promised @maximusunc  (cc: @sierra-moxon), I will try to express my sense of the testing landscape. I suspect there is no "right answer" to this... that your current implementation is equally tenable...

Anyhow, I would see the planning of the test process as an ordered series of test configuration decisions. Here's an order I have sort of followed in earlier testing work:

1.  which component do you wish to test (UI, ARS, ARA, KP, ??)
2.  within which environment are you running the component to be tested?
3.  which category of test ("TestRunner") do you which to run against the component
4.  within what global context do you wish to test (e.g. TRAPI version, Biolink version, ???)
5.  what test data (TestAssets/Cases/Suites) do you wish to use as input?
6.  when do you want to run the test?
7.  how do you wish to assess the test results?

My general sense is that the answers to all of the above questions coudl be aggregated within a top level **TestRunSession** declaration.

What I understand, however, is that from your perspective, a **TestRunSession** is mainly a scheduled (timestamped) container for any _ad hoc_ set of test data (TestAssets/Cases/Suites) to be processed, and that the test data itself will specify where it is consumed (i.e. which component) thus different members of the **`test_entities`** array of test data could be routed to different target components, albeit, all processed by the same specified **TestRunner**.

Nothing wrong with that, I guess... one is simply embedding the answer to question 1 above, inside the test data returned in response to question 5.

Just an alternate way of organizing the testing... "cutting the cake..." LOL.

**Review by sierra-moxon — APPROVED** [review]

File: src/translator_testing_model/datamodel/pydanticmodel.py (hunk 5 above)
Comment (sierra-moxon) [reply src/translator_testing_model/datamodel/pydanticmodel.py:493]: I think we use TestCase for the runner to help collate assets into a single query.  Happy to talk this through on a call.

Comment (maximusunc) [reply src/translator_testing_model/datamodel/pydanticmodel.py:493]: I'm fine with removing these subclass Suite classes. In the Test Harness, I basically start with grabbing a single TestSuite, and then for each TestCase, I look at the specified test_case_objective to determine which Runner to send it to.
I would also argue that a Suite can have different types of TestCases within it, so these subclass Suites might not even actually fit well.

**Pull request merged**